numpy/core/arrayprint.py

```diff
@@ -507,14 +530,9 @@ def array2string(a, max_line_width=None, precision=None,
         formatter = _formatter
 
     if a.shape == ():
-        x = a.item()
-        if a.dtype.fields is not None:
-            arr = array([x], dtype=a.dtype)
-            format_function = _get_format_function(
-                    arr, precision, suppress_small, formatter)
-            lst = format_function(arr[0])
-        else:
-            lst = style(x)
+        format_function = _get_format_function(
+                a.reshape((1,)), precision, suppress_small, formatter)
+        lst = format_function(a[()])
```

Comment:
OK I looked a bit more at this, and I remember another problem that makes me want to keep it this way: There is an cyclic import problem that seems complicated to fix.

Below is a traceback if I remove this special case which occurs when I try to import numpy.

The problem is that numpy `core` modules (specifically `getlimits`) want to call `array2string` before all numpy submodules are loaded. However `array2string` needs to be able to call `np.ravel` which checks for `np.matrix`, which hasn't been loaded into numpy at that point, so fails.

I don't see any simple fix to this involving changing the import order of the submodules, since `matrix` itself depends on `getlimits`, so there is a true cyclic import problem there.

This all sounds like something I'd rather work around in this PR. Perhaps the fix to to make `getlimits` not call `array2string`, or perhaps `array2string` not depend on `ravel`, I don't know, but in any case it seems fairly unrelated to 0d reprs.


```
Traceback (most recent call last):
  File "runtests.py", line 469, in <module>
    main(argv=sys.argv[1:])
  File "runtests.py", line 179, in main
    import numpy as np
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/__init__.py", line 142, in <module>
    from . import add_newdocs
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/add_newdocs.py", line 13, in <module>
    from numpy.lib import add_newdoc
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/lib/__init__.py", line 8, in <module>
    from .type_check import *
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/lib/type_check.py", line 11, in <module>
    import numpy.core.numeric as _nx
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/__init__.py", line 51, in <module>
    from . import getlimits
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/getlimits.py", line 163, in <module>
    tiny=_tiny_f128)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/getlimits.py", line 82, in __init__
    self._str_eps = float_to_str(self.eps)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/getlimits.py", line 71, in <lambda>
    float_to_str = lambda v: (params['fmt'] % array(_fr0(v)[0], ftype))
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/numeric.py", line 1939, in array_str
    return array2string(a, max_line_width, precision, suppress_small, ' ', "")
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/arrayprint.py", line 409, in wrapper
    return f(self, *args, **kwargs)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/arrayprint.py", line 537, in array2string
    separator, prefix, formatter=formatter)
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/arrayprint.py", line 372, in _array2string
    data = ravel(asarray(a))
  File "/home/allan/nmpy_dev/build/testenv/lib/python2.7/site-packages/numpy/core/fromnumeric.py", line 1481, in ravel
    if isinstance(a, np.matrix):
AttributeError: 'module' object has no attribute 'matrix'
```